src/memory.rs

```diff
@@ -0,0 +1,63 @@
+use std::cell::Cell;
+use rand::rngs::StdRng;
+
+use rustc_mir::interpret::{Pointer, Allocation, AllocationExtra, InterpResult};
+use rustc_target::abi::Size;
+
+use crate::{stacked_borrows, intptrcast};
+use crate::stacked_borrows::{Tag, AccessKind};
+
+#[derive(Default, Clone, Debug)]
+pub struct MemoryState {
+    pub stacked: stacked_borrows::MemoryState,
+    pub intptrcast: intptrcast::MemoryState,
+    /// The random number generator to use if Miri
+    /// is running in non-deterministic mode
+    pub(crate) rng: Option<StdRng>
+}
+
+#[derive(Debug, Clone,)]
+pub struct AllocExtra {
+    pub stacks: stacked_borrows::Stacks,
+    pub base_addr: Cell<Option<u64>>,
+}
+
+impl AllocationExtra<Tag> for AllocExtra {
+    #[inline(always)]
+    fn memory_read<'tcx>(
+        alloc: &Allocation<Tag, AllocExtra>,
+        ptr: Pointer<Tag>,
+        size: Size,
+    ) -> InterpResult<'tcx> {
+        trace!("read access with tag {:?}: {:?}, size {}", ptr.tag, ptr.erase_tag(), size.bytes());
+        alloc.extra.stacks.for_each(ptr, size, |stack, global| {
+            stack.access(AccessKind::Read, ptr.tag, global)?;
+            Ok(())
+        })
+    }
```

Comment:
There shouldn't be any Stacked Borrows logic here. Please make these methods just forward to methods with the same name on `stacks`.